### PR TITLE
1856 nationalization core

### PR DESCRIPTION
**Pull request by ventusignis**
This isn't everything for nationalization but this does do a good chunk of it and the goal is that I'll be able to add on extra functionality later with small PRs

What this does do is:
- Nationalization starts with the first 6T
- Player order starts with the player who bought the 6T
- Companies are forced to repay loans
- Presidents of companies, in player order, choose to cover the difference to payoff loans or let the company merge in
- Companies when they merge transfer leftover cash from paying off loans and trains
- Tokens of merged corporations are given to the national
- Trains of merged corporations are given to the national
- Shares of merged corporations are traded 2-1 for national shares
- Market price of national is the average* of the merged corporations prices

### CODE REVIEW

File: lib/engine/g_1856/corporation.rb [hunk 1]
```diff
@@ -16,6 +17,7 @@ def initialize(game, sym:, name:, **opts)
         @started = false
         @escrow = nil
         super(sym: sym, name: name, **opts)
+        @all_shares = shares_by_corporation[self].dup
         @capitalization = nil
       end
```
Comment (ventusignis) [inline lib/engine/g_1856/corporation.rb:23]: The floating behavior in 1856 is going to keep the 1856::Corporation class around although escrow logic can probably be pulled into the base classes at some point which could reduce the amount of custom 1856 code

File: lib/engine/ability/unparrable.rb [hunk 2]
```diff
@@ -0,0 +1,10 @@
+# frozen_string_literal: true
+
+require_relative 'base'
+
+module Engine
```
Comment (tobymao) [inline lib/engine/ability/unparrable.rb:5]: why is this needed, can't you just define can_par? in game like others?

File: lib/engine/g_1856/corporation.rb (hunk 1 above)
Comment (tobymao) [reply lib/engine/g_1856/corporation.rb:23]: what about it?

File: lib/engine/game/g_1856.rb [hunk 3]
```diff
@@ -355,6 +385,346 @@ def stock_round
...
+          @loans << corp.loans.pop(loans_repaid)
+          @log << "#{corp.name} repays #{format_currency(amount_repaid)} to redeem #{loans_repaid} loans"
+        end
+      end
+
+      def merge_major(major)
+        @national_formed = true
+        @log << "-- #{major.name} merges into #{national.name} --"
+        # Trains are transferred
+        major.trains.clone.each do |t|
```
Comment (tobymao) [inline lib/engine/game/g_1856.rb:432]: dup

File: lib/engine/game/g_1856.rb [hunk 4]
```diff
@@ -355,6 +385,346 @@ def stock_round
...
+        home_bases = @nationalized_corps.map do |c|
+          nationalize_home_token(c, create_national_token)
+        end
+        # So the national will get 11 tokens if and only if all 11 majors merge in
+        remaining_tokens = [national_token_limit - home_bases.size, 0].max
+
+        # Other tokens second, ignoring duplicates from the home token set
+        @nationalized_corps.each do |corp|
+          corp.tokens.each do |token|
+            next if token.city.nil? || home_bases.include?(token.city.hex)
```
Comment (tobymao) [inline lib/engine/game/g_1856.rb:620]: !token.city

File: lib/engine/game/g_1856.rb [hunk 5]
```diff
@@ -355,6 +385,346 @@ def stock_round
...
+        @national.tokens.each do |token|
+          city = token.city
+          token.remove! if cities.include?(city)
+        end
+      end
+
+      # Convert the home token of the corporation to one of the national's
+      # Return the nationalized corps home hex
+      def nationalize_home_token(corp, token)
+        if token.nil?
```
Comment (tobymao) [inline lib/engine/game/g_1856.rb:683]: unless token

File: lib/engine/game/g_1856.rb (hunk 3 above)
Comment (ventusignis) [reply lib/engine/game/g_1856.rb:432]: Is `dup` generally preferred to `clone`, or just in this case?

Comment (tobymao) [reply lib/engine/game/g_1856.rb:432]: dup is just a soft copy, clone does all internals which is not necessary

File: lib/engine/game/g_1856.rb (hunk 4 above)
Comment (ventusignis) [reply lib/engine/game/g_1856.rb]: Changed

File: lib/engine/game/g_1856.rb (hunk 3 above)
Comment (ventusignis) [reply lib/engine/game/g_1856.rb]: OK thanks, I've changed this

File: lib/engine/game/g_1856.rb (hunk 5 above)
Comment (ventusignis) [reply lib/engine/game/g_1856.rb]: Changed

File: lib/engine/ability/unparrable.rb (hunk 2 above)
Comment (ventusignis) [reply lib/engine/ability/unparrable.rb]: I did some changes to remove this ability; I also defined `float_str` which will communicate the same information to the player
image

File: lib/engine/g_1856/corporation.rb (hunk 1 above)
Comment (ventusignis) [reply lib/engine/g_1856/corporation.rb:23]: You've mentioned a desire to put the custom logic in `G1856::Coporation` into the 1856 game file before. I have some ideas on how to incrementally remove some of the custom classes for 1856 but I think that the floating behavior is going to keep the custom 1856 corporation around for that purpose

File: lib/engine/g_1856/corporation.rb [hunk 6]
```diff
@@ -16,6 +17,7 @@ def initialize(game, sym:, name:, **opts)
         @started = false
         @escrow = nil
         super(sym: sym, name: name, **opts)
+        @all_shares = shares_by_corporation[self].dup
```
Comment (tobymao) [inline lib/engine/g_1856/corporation.rb:20]: is this really necesasry? i see it's for total_shares, but can't your 1856 game class count this without having to dup the shares?

File: lib/engine/g_1856/corporation.rb [hunk 7]
```diff
@@ -45,11 +47,8 @@ def capitalization_type_desc
         CAPITALIZATION_STRS[@capitalization || _capitalization_type]
       end
 
-      def release_escrow!
-        @log << "Releasing #{@game.format_currency(@escrow)} from escrow for ${@name}"
-        @cash += @escrow
-        @escrow = nil
-        @capitalization = :incremental
+      def total_shares
```
Comment (tobymao) [inline lib/engine/g_1856/corporation.rb:50]: this doesn't seem like it needs to be in corporation class

File: lib/engine/g_1856/corporation.rb (hunk 6 above)
Comment (ventusignis) [reply lib/engine/g_1856/corporation.rb:20]: It's also used in `Engine::Game::G1856::national_issue_shares!` for converting the existing shares to 5% shares and for accessing the shares in `Engine::Game::G1856::national_share_swap` for distributing shares among players. I also have to solve how to give a player a 1-share presidents cert in one of the nationalization edge cases, but that's not solved yet

File: lib/engine/g_1856/corporation.rb (hunk 7 above)
Comment (ventusignis) [reply lib/engine/g_1856/corporation.rb:50]: This is overriding `total_shares` from the base Corporation class

Comment (tobymao) [reply lib/engine/g_1856/corporation.rb:50]: what was wrong with the initial implementation?

Comment (ventusignis) [reply lib/engine/g_1856/corporation.rb:50]: It helps with some of the weird edge cases in 1856 but since I cut off this PR before I started to tackle them I'm going to remove that from here

**Review by tobymao — APPROVED** [review]
